api/oc_ri_server_internal.h

```diff
@@ -0,0 +1,111 @@
+/****************************************************************************
+ *
+ * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2023 plgd.dev s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"),
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************************/
+
+#ifndef OC_RI_SERVER_INTERNAL_H
+#define OC_RI_SERVER_INTERNAL_H
+
+#include "oc_ri.h"
+#include "util/oc_compiler.h"
+
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief allocate a resource structure
+ *
+ * @return oc_resource_t*
+ */
+oc_resource_t *oc_ri_alloc_resource(void);
+
+/**
+ * @brief deallocate a resource structure
+ *
+ * @param resource the resource to be deallocated
+ */
+void oc_ri_dealloc_resource(oc_resource_t *resource);
+
+/**
+ * @brief Callback invoked on resource before it is deleted by
+ * oc_delayed_delete_resource.
+ *
+ * @param resource Resource to be deleted
+ * @param data Custom user data provided to the callback
+ */
+typedef void (*oc_ri_delete_resource_cb_t)(oc_resource_t *resource, void *data);
+
+typedef struct oc_ri_on_delete_resource_t
+{
+  struct oc_ri_on_delete_resource_t *next; ///< next item
+  oc_ri_delete_resource_cb_t cb;           ///< callback to be invoked
+  void *data;                              ///< data for the callback
```

Comment:
data is not needed -> I know this is pretty but we are in embedded world and user need to have stored the object in his app.